src/Functions/FunctionsStringArray.h

```diff
@@ -263,6 +263,89 @@ class SplitByStringImpl
     }
 };
 
+class SplitByRegexpImpl
+{
+private:
+    Regexps::Pool::Pointer re;
+    OptimizedRegularExpression::MatchVec matches;
+
+    Pos pos;
+    Pos end;
+public:
+    static constexpr auto name = "splitByRegexp";
+    static String getName() { return name; }
+    static size_t getNumberOfArguments() { return 2; }
+
+    /// Check the type of function arguments.
+    static void checkArguments(const DataTypes & arguments)
+    {
+        SplitByStringImpl::checkArguments(arguments);
+    }
+
+    /// Initialize by the function arguments.
+    void init(const ColumnsWithTypeAndName & arguments)
+    {
+        const ColumnConst * col = checkAndGetColumnConstStringOrFixedString(arguments[0].column.get());
+
+        if (!col)
+            throw Exception("Illegal column " + arguments[0].column->getName()
+                            + " of first argument of function " + getName() + ". Must be constant string.",
+                            ErrorCodes::ILLEGAL_COLUMN);
+
+        if (!col->getValue<String>().empty())
+            re = Regexps::get<false, false>(col->getValue<String>());
+
+        matches.resize(1);
+    }
+
+    /// Returns the position of the argument that is the column of strings
+    size_t getStringsArgumentPosition()
+    {
+        return 1;
+    }
+
+    /// Called for each next string.
+    void set(Pos pos_, Pos end_)
+    {
+        pos = pos_;
+        end = end_;
+    }
+
+    /// Get the next token, if any, or return false.
+    bool get(Pos & token_begin, Pos & token_end)
+    {
+        if (!re)
```

Comment:
Query `SELECT splitByRegexp('x*', 'xxx123xxx')` returns `['','123xxx']`, looks a bit weird. But I'm not sure how it should be. Maybe we can test if empty string match regex and consider it as empty regex. Or maybe it's useless corner-case and current behavior appropriate.

Reply:
This logic is similar with extractAll.
I tried ` SELECT extractAll( 'xxx123xxx', 'x*');` returns `['xxx'] `. These two functions share the same logic.

Reply:
Ok, let's leave it as is